packages/doenetml-worker-rust/lib-doenetml-core/src/core/dast/macro_resolve.rs

```diff
@@ -0,0 +1,230 @@
+//! Macros use a `path`, which consists of _names_ separated by dots (e.g. `a.b.c`), possibly followed by
+//! state variable names separated by dots (e.g. `a.b.c.x.y.z`). A `Resolver` searches for a matching node
+//! from a given starting position. Because a `Resolver` does not know whether part of a path corresponds to a
+//! name or state variable name, the longest valid partial match is used and unmatched portions of a `path` are
+//! preserved for future use.
+
+use std::{collections::HashMap, iter, mem};
+
+use crate::dast::flat_dast::{FlatNode, UntaggedContent};
+
+use super::{
+    flat_dast::{FlatRoot, Index},
+    PathPart,
+};
+use serde::Serialize;
+use thiserror::Error;
+
+#[derive(Clone, Debug, Error, PartialEq)]
+pub enum ResolutionError {
+    #[error("No node identified by path")]
+    NoReferent,
+    #[error("Path referred to more than one node")]
+    NonUniqueReferent,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct RefResolution {
+    pub node_idx: Index,
+    pub unresolved_path: Option<Vec<PathPart>>,
+}
+
+/// Status of a pointer referring to children of an element.
+#[derive(Debug)]
+enum Ref {
+    Unique(Index),
+    Ambiguous(Vec<Index>),
+}
+
+/// A `Resolver` is used to lookup elements by path/name. It constructs a search index
+/// upon construction. If the underlying `FlatRoot` changes, a new `Resolver` should be
+/// recreated.
+#[derive(Debug)]
+pub struct Resolver {
+    /// List of the parent of a node at a given index.
+    node_parent: Vec<Option<Index>>,
+    /// Map of all the names that are accessible (as descendants) from a given node.
+    name_map: Vec<HashMap<String, Ref>>,
+}
+
+impl Resolver {
+    pub fn from_flat_root(flat_root: &FlatRoot) -> Self {
+        Resolver {
+            node_parent: flat_root.nodes.iter().map(|node| node.parent()).collect(),
+            name_map: Self::build_name_map(flat_root),
+        }
+    }
+
+    /// Search for a node specified by `path` starting from `origin`. This algorithm searches first
+    /// for the nearest parent that has an element with `path[0].name` as its name. Then it tries
```

Comment:
I was confused by the phrase "has an element with". I first read that the parent element had to have that name, which I know isn't correct. Maybe "has a descendant element with"?